src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java

```diff
@@ -217,106 +219,142 @@ public StorageCalculator(boolean forArguments, boolean forVariadicFunction) {
+        Windows, non-VF | CW in regs       | CW on the stack                 | CW on the stack
+        Windows, VF     | CW in regs       | CW split between regs and stack | CW on the stack
+         */
+        StructStorage[] structStorages(GroupLayout layout, boolean forHFA) {
+            int regType = forHFA ? StorageType.VECTOR : StorageType.INTEGER;
+            int numChunks = (int)Utils.alignUp(layout.byteSize(), MAX_COPY_SIZE) / MAX_COPY_SIZE;
+            int requiredStorages = forHFA ? layout.memberLayouts().size() : numChunks;
+            boolean hasEnoughRegisters = hasEnoughRegisters(regType, requiredStorages);
+
+            // For the ABI variants that pack arguments spilled to the
+            // stack, HFA arguments are spilled as if their individual
+            // fields had been allocated separately rather than as if the
+            // struct had been spilled as a whole.
+            boolean useFieldWiseSpill = requiresSubSlotStackPacking() && !forVarArgs;
+            boolean isFieldWise = forHFA && (hasEnoughRegisters || useFieldWiseSpill);
+            if (!isFieldWise) {
+                requiredStorages = numChunks;
+            }
+
+            boolean spillPartially = forVariadicFunction && spillsVariadicStructsPartially();
+            boolean furtherAllocationFromTheStack = !hasEnoughRegisters && !spillPartially;
+            if (furtherAllocationFromTheStack) {
                 // Any further allocations for this register type must
                 // be from the stack.
-                nRegs[type] = MAX_REGISTER_ARGUMENTS;
-                return null;
+                nRegs[regType] = MAX_REGISTER_ARGUMENTS;
             }
-        }
 
-        VMStorage[] regAlloc(int type, MemoryLayout layout) {
-            boolean spillRegistersPartially = forVariadicFunction && spillsVariadicStructsPartially();
+            if (requiresSubSlotStackPacking() && !isFieldWise) {
+                // Pad to the next stack slot boundary instead of packing
+                // additional arguments into the unused space.
+                alignStack(STACK_SLOT_SIZE);
+            }
 
-            return spillRegistersPartially ?
-                regAllocPartial(type, layout) :
-                regAlloc(type, requiredRegisters(layout));
-        }
+            StructStorage[] structStorages = new StructStorage[requiredStorages];
+            long offset = 0;
+            for (int i = 0; i < structStorages.length; i++) {
+                ValueLayout copyLayout;
+                if (isFieldWise) {
+                    // fields only, no padding
+                    copyLayout = (ValueLayout) layout.memberLayouts().get(i);
+                } else {
+                    // chunk-wise copy
+                    long copySize = Math.min(layout.byteSize() - offset, MAX_COPY_SIZE);
+                    boolean useFloat = false; // never use float for chunk-wise copies
```

Comment:
this local seems unnecessary?

Reply:
The local is there to give a name to the magic `false` value, and it also allows putting a comment. I preferred doing this to the alternative:

    copyLayout = SharedUtils.primitiveLayoutForSize(copySize, /* useFloat = */ false); // never use float for chunk-wise copies